types/wlr_data_control_v1.c

```diff
@@ -105,18 +81,144 @@ static const struct zwlr_data_control_source_v1_interface source_impl = {
 	.destroy = source_handle_destroy,
 };
 
+static void data_control_source_destroy(struct data_control_source* source) {
+	char **p;
+	wl_array_for_each(p, &source->mime_types) {
+		free(*p);
+	}
+	wl_array_release(&source->mime_types);
+
+	if (source->active_source != NULL) {
+		wlr_data_source_destroy(source->active_source);
+	} else if (source->active_primary_source != NULL) {
+		wlr_primary_selection_source_destroy(
+			source->active_primary_source);
+	}
+
+	free(source);
+}
+
 static void source_handle_resource_destroy(struct wl_resource *resource) {
-	struct client_data_source *source = source_from_resource(resource);
+	struct data_control_source *source = source_from_resource(resource);
+	if (source == NULL) {
+		return;
+	}
+
+	// Prevent destructors called from data_control_source_destroy() from
+	// accessing invalid data.
+	wl_resource_set_user_data(resource, NULL);
```

Comment:
This should be done in `data_control_source_destroy`. See the CONTRIBUTING guide about inert resources.